### PR TITLE
cpu/cc2538: Add RF driver

### PR DESCRIPTION
**Pull request by aeneby**
This PR constitutes a working RF driver for the TI CC2538 MCU. Although not yet feature-complete, both sending and receiving of packets works correctly through the gnrc/netdev2 stack. My hope is that it is complete enough to be considered for inclusion, pending further testing and review.

### CODE REVIEW

**Comment by aeneby** [comment]
@anonmall @hexluthor @kYc0o Would be good to get this tested on cc2538dk too, I've only tested on OpenMote so far (not that there should be any significant differences). Feedback welcome.

**Comment by PeterKietzmann** [comment]
@aeneby do I see it correctly that the radio is the same on the [remote](http://zolertia.io/product/hardware/re-mote)? In that case, could you add support for it as well?

**Comment by kYc0o** [comment]
Sorry for the absence, I can test this and try to merge it for the Hack'n'ACK tomorrow on the cc2538dk.

**Comment by aeneby** [comment]
@PeterKietzmann yes looks to be the same MCU. I'll have a look.

@kYc0o thanks. I'd like to see #5297 merged at some point too, which will improve the functionality of this driver (and others). So if you get time at the Hack'n'ACK tomorrow please take a look :)

**Comment by aeneby** [comment]
@PeterKietzmann I've committed changes to support RE-Mote as well, hopefully - I don't have the hardware to test, but it seems to build.

**Comment by aeneby** [comment]
@kYc0o I just realised that if you want to test on the cc2538dk you'll need to enable the driver for that platform too, so I have included a commit for that.

**Comment by PeterKietzmann** [comment]
Without looking at the code I just wanted to report that I was able to use the driver with a remote (updated revision A). In addition I was able to ping around and sniff with a Raspi+openlabs transceiver.

```
 > ifconfig
 Iface  7   HWaddr: ef:7d  Channel: 17  Page: 0  NID: 0x4fff
            Long HWaddr: ff:fb:0b:bf:bf:ff:bf:ef 
            TX-Power: 3dBm  State: IDLE 
            AUTOACK  MTU:1280  HL:64  
            Source address length: 2
            Link type: wireless
            inet6 addr: ff02::1/128  scope: local [multicast]
            inet6 addr: fe80::ff:fe00:ef7d/64  scope: local
            inet6 addr: ff02::1:ff00:ef7d/128  scope: local [multicast
```
Are the addresses set as intended? In RIOT we usually have RF channel `26` and PAN `0x0023` as default.

**Comment by alignan** [comment]
````
2016-05-31 13:59:35,680 - INFO # > ifconfig
2016-05-31 13:59:35,682 - INFO # Iface  4   HWaddr: b9:0f  Channel: 17  Page: 0  NID: 0xabcd
2016-05-31 13:59:35,682 - INFO #            Long HWaddr: b9:0f:16:06:00:4b:12:00 
2016-05-31 13:59:35,683 - INFO #            TX-Power: 3dBm  State: IDLE 
2016-05-31 13:59:35,683 - INFO #            AUTOACK  
2016-05-31 13:59:35,684 - INFO #            Source address length: 2
````

I can confirm it is working for the `RE-Mote revision A`, I'm updating #4748 to reflect this new contribution, thanks!

**Comment by kYc0o** [comment]
@PeterKietzmann how did you succeeded on pinging? I'm trying with the remote and the address seems a bit strange:

```
2016-05-31 20:12:25,741 - INFO # > ifconfig
2016-05-31 20:12:25,743 - INFO # Iface  7   HWaddr: ff:9f  Channel: 26  Page: 0  NID: 0x23
2016-05-31 20:12:25,744 - INFO #            Long HWaddr: bf:ac:da:d7:b3:fe:bb:fc 
2016-05-31 20:12:25,744 - INFO #            TX-Power: 3dBm  State: IDLE 
2016-05-31 20:12:25,745 - INFO #            AUTOACK  MTU:1280  HL:64  
2016-05-31 20:12:25,746 - INFO #            Source address length: 2
2016-05-31 20:12:25,746 - INFO #            Link type: wireless
2016-05-31 20:12:25,747 - INFO #            inet6 addr: ff02::1/128  scope: local [multicast]
2016-05-31 20:12:25,748 - INFO #            inet6 addr: fe80::ff:fe00:ff9f/64  scope: local
2016-05-31 20:12:25,749 - INFO #            inet6 addr: ff02::1:ff00:ff9f/128  scope: local [multicast]
2016-05-31 20:12:25,749 - INFO #            
2016-05-31 20:12:25,749 - INFO #            Statistics for Layer 2
2016-05-31 20:12:25,750 - INFO #             RX packets 0  bytes 0
2016-05-31 20:12:25,751 - INFO #             TX packets 0 (Multicast: 0)  bytes 0
2016-05-31 20:12:25,751 - INFO #             TX succeeded 0 errors 0
2016-05-31 20:12:25,752 - INFO #            Statistics for IPv6
2016-05-31 20:12:25,752 - INFO #             RX packets 0  bytes 0
2016-05-31 20:12:25,753 - INFO #             TX packets 0 (Multicast: 0)  bytes 0
2016-05-31 20:12:25,753 - INFO #             TX succeeded 0 errors 0
2016-05-31 20:12:25,753 - INFO # 
```

**Comment by PeterKietzmann** [comment]
Yes as I said, address generation seems a bit weird to me. I didn't express myself correctly. Didn't ping between nodes *but* sniffed the ICMPv6 message and the parameters looked as expected

**Comment by kYc0o** [comment]
Ok I see. I did a test on the Re-Mote Rev A with the default example and it is able to send packets but the receiving I think is broken. I didn't investigate why but it can be time consuming...

**Comment by PeterKietzmann** [comment]
Are you at it? Did you cross-check with an other board? E.g. Atmel, iotlab or phyNODE?

**Comment by kYc0o** [comment]
Yes I tested with samr21, it receives the packets sent by the remote but the remote does not receive packets from samr21

File: cpu/cc2538/include/cc2538_rf.h [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+#define IEEE802154_CHANNEL_SPACING  (5)    /**< Channel spacing in MHz  */
+
+#define IEEE802154_CHAN2FREQ(chan)  ( IEEE802154_MIN_FREQ + ((chan) - IEEE802154_MIN_CHANNEL) * IEEE802154_CHANNEL_SPACING )
+#define IEEE802154_FREQ2CHAN(freq)  ( IEEE802154_MIN_CHANNEL + ((freq) - IEEE802154_MIN_FREQ) / IEEE802154_CHANNEL_SPACING )
+/* /TODO */
+
+#define CC2538_MIN_FREQ             (2394)
+#define CC2538_MAX_FREQ             (2507)
+
+#define CC2538_RF_POWER_DEFAULT     + 3 /**< Default output power in dBm */
```
Comment (kYc0o) [inline cpu/cc2538/include/cc2538_rf.h]: Why are you specifying that is positive? Can you put it in parenthesis?

File: cpu/cc2538/include/cc2538_rf.h [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+#define CC2538_RF_CHANNEL_DEFAULT   17
+
+#define OUTPUT_POWER_MIN            -24  /**< Min output power in dBm */
+#define OUTPUT_POWER_MAX            + 7  /**< Max output power in dBm */
```
Comment (kYc0o) [inline cpu/cc2538/include/cc2538_rf.h]: Same here...

File: cpu/cc2538/radio/cc2538_rf_netdev.c [hunk 3]
```diff
@@ -0,0 +1,379 @@
...
+const netdev2_driver_t cc2538_rf_driver = {
+    .get  = _get,
+    .set  = _set,
+    .send = _send,
+    .recv = _recv,
+    .isr  = _isr,
+    .init = _init,
+};
+
+// Reference pointer for the IRQ handler
```
Comment (kYc0o) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: please use C style comments

File: cpu/cc2538/radio/cc2538_rf_netdev.c [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+{
+    cc2538_rf_t *dev = (cc2538_rf_t *)netdev;
+
+    if (dev == NULL) {
+        return -ENODEV;
+    }
+
+    switch (opt) {
+        // NOTE: Handling of NETOPT_AUTOACK in netdev2_ieee802154
+        // is broken, so do it here instead. See PR #5297.
```
Comment (kYc0o) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: same here

File: cpu/cc2538/radio/cc2538_rf_netdev.c [hunk 5]
```diff
@@ -0,0 +1,379 @@
...
+            else {
+                *((netopt_enable_t *)value) = NETOPT_DISABLE;
+            }
+            return sizeof(netopt_enable_t);
+
+        case NETOPT_CHANNEL_PAGE:
+            if (max_len < sizeof(uint16_t)) {
+                return -EOVERFLOW;
+            }
+            // This tranceiver only supports page 0
```
Comment (kYc0o) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: and again :)

File: cpu/cc2538/radio/cc2538_rf_netdev.c [hunk 6]
```diff
@@ -0,0 +1,379 @@
...
+#include <errno.h>
+
+#include "net/gnrc.h"
+#include "net/netdev2.h"
+
+#include "cc2538_rf.h"
+#include "cc2538_rf_netdev.h"
+#include "cc2538_rf_internal.h"
+
+#define ENABLE_DEBUG 1
```
Comment (kYc0o) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: Please disable debug, even if there is only one `DEBUG` macro on this file.

**Comment by alignan** [comment]
> Ok I see. I did a test on the Re-Mote Rev A with the default example and it is able to send packets but the receiving I think is broken. I didn't investigate why but it can be time consuming...

@kYc0o Did you see my comment above? I was able to send and receive data using 2 `RE-Motes` (revision A).

**Comment by kYc0o** [comment]
Did you tested it with other boards? My setup is a Re-Mote and a samr21-xpro. I'm trying to debug it but I cannot find some of the receiver and interrupt functions...

**Comment by alignan** [comment]
No I haven't, I could use an `OpenMote` or a 'cc2538dk', but I think result would be pretty much the same... I will try to get my hands on other type of board.

**Comment by aeneby** [comment]
@kYc0o thanks, I think I've addressed your comments with the latest commits.

> In RIOT we usually have RF channel 26 and PAN 0x0023 as default.

@PeterKietzmann I wasn't aware there was a RIOT default - no, the channel and PAN are not set explicitly, but rather it uses the cc2538's default values. I vaguely remember reading in the cc2538 user guide that the default HW address is generated based on an ID unique to each MCU, for example.

**Comment by kYc0o** [comment]
I was experimenting packet receiving from an accidentally powered BR (sending broadcast packets), thus some packets were printed by the default example... Maybe an addressing issue? I'll try to give more feedback on this.

Can you rebase since #5297 was merged? This way I can test the overall functionality.

**Comment by aeneby** [comment]
Yup, rebased and squashed.

Just an aside, the RPi 2 I was using for packet sniffing has given up on me. I have an RPi 3 but I'm struggling to get it working with the Openlabs at86rf233 radio. Has anyone managed to get this radio working with the RPi 3? I can't see any reason why it shouldn't work...

**Comment by miri64** [comment]
Needs another rebase and adaption to https://github.com/RIOT-OS/RIOT/pull/4871.

**Comment by aeneby** [comment]
Rebased and adapted.

**Comment by kYc0o** [comment]
Can you try to rebase again? It seems it's conflicting...

**Comment by aeneby** [comment]
Rebased.

**Comment by aeneby** [comment]
Rebased to include #5523 (should help with testing)

**Comment by alignan** [comment]
Is there anything else pending to have this merged? happy to help.

**Comment by kYc0o** [comment]
I tried this PR with cc2538dk and I was able to receive only broadcasts, when sending packets to the L2 (long) address it didn't work, nor the ipv6 address... @alignan can you test unicast messages?

**Comment by aeneby** [comment]
I've just done a quick unicast test using short L2 addresses - there's some kind of little/big endian mixup here. For example, sending to `11:22` works, but only if the receiving node's short address is configured to `22:11`.

**Comment by aeneby** [comment]
New commit to fix byte order in address getting/setting, and rebased to current master.

@kYc0o are you able to re-test?

**Comment by kYc0o** [comment]
Yes I'll try again ASAP

**Comment by aeneby** [comment]
Rebased again to current master.

**Comment by emmanuelsearch** [comment]
@aeneby what's the status of this PR, is it ready from your point of view?

**Comment by aeneby** [comment]
@emmanuelsearch yes, just waiting for @kYc0o.

**Comment by aeneby** [comment]
Rebased, squashed, and removed `[ci skip]`.

**Comment by kYc0o** [comment]
Hi again! Sorry for the loooong delay, I was a bit busy with some other stuff so no time to look into this.

I took a look to the code and there are several things  to do, especially to be compatible with netdev2.

One first think is that you have squashed all the things into one commit, while you're affecting more than a part of RIOT. It means that if you added support for this new feature to cc2538dk board, it's better to do it in a separated commit, as well as for the other boards supporting it.
So, for instance, you can split on:

- cpu/cc2538: ...
- boards/cc2538dk: ...
- boards/openmote-cc2538: ...
- boards/remote-*: (however there is a commit for this too in #5587 )

and maybe change the PR message to cpu/cc2538: add RF driver or something similar.

For netdev2, usually we pass the `netdev2_t` object to any function that changes the actual registers on the device, e.g. channel, sensibility, TX power, addresses (short and long), as well as in the init function, in order to save them in this object, which is used for netstats and other tools to get information about the device.
You can take a look into the at86rf2xx or cc2420 drivers, as a source of inspiration.

**Comment by alignan** [comment]
If you update your commit with #5587 I would be more than OK to close it, as the ``RE-Mote` platform updates has been merged already

**Comment by aeneby** [comment]
Welcome back @kYc0o :)

> One first think is that you have squashed all the things into one commit

Fixed; also rebased to current master.

> and maybe change the PR message

Done.

> For netdev2, usually we pass the netdev2_t object to any function that changes the actual registers on the device, e.g. channel, sensibility, TX power, addresses (short and long)

But isn't this already being done, using `netdev2_ieee802154_{g,s}et()`? See [here](https://github.com/RIOT-OS/RIOT/pull/5291/files#diff-b8d088b7424905055347a5d1f7bd781cR247), for example.

> which is used for netstats

No, netstats hasn't been implemented in this driver yet. But is it really a blocker for this PR? Since it's defined as a module it seems pretty independent from the rest of the code, and could be implemented in a separate PR.

> If you update your commit with #5587 I would be more than OK to close it

@alignan I have tried to include the relevant commits, please double check that everything looks correct though before closing your PR.

**Comment by kYc0o** [comment]
> But isn't this already being done, using netdev2_ieee802154_{g,s}et()? See here, for example.
Yes, you're right, it's only that in your init function I can't see how to setup the default netdev parameters, especially for this:

```
    /* set default protocol */
#ifdef MODULE_GNRC_SIXLOWPAN
    dev->netdev.proto = GNRC_NETTYPE_SIXLOWPAN;
#elif MODULE_GNRC
    dev->netdev.proto = GNRC_NETTYPE_UNDEF;
#endif
```
which are used to tell the driver how to manage the received packets.

Did you put this in another function maybe?

**Comment by kYc0o** [comment]
Can you rebase since #5615 was merged?

**Comment by kYc0o** [comment]
I can confirm L2 messages working, either with short and long addresses, but IPv6 is not working since the protocol is not specified.

**Comment by aeneby** [comment]
> Did you put this in another function maybe?

I've now added setting of `netdev.proto` in the netdev2 driver init function, so please retry. I've also rebased to current master.

**Comment by kYc0o** [comment]
Excellent! I tested it in a remote-reva with success! However, the NID (PAN ID) it's set randomly... can you check that?

**Comment by alignan** [comment]
> I have tried to include the relevant commits, please double check that everything looks correct though before closing your PR.

@aeneby thanks for including the changes, everything looks OK.

**Comment by aeneby** [comment]
@kYc0o, as per @PeterKietzmann's [comment](https://github.com/RIOT-OS/RIOT/pull/5291#issuecomment-222658909), I've set the default PAN to `0x0023` and the default channel to 26.

**Comment by PeterKietzmann** [comment]
Is it intended that hardware addresses (and derived by these also IPv6 addresses) are non-volatile? Usually we intend to have the "same" address on one node after initialisation. However, there are still open topics about the default address in #5457

**Comment by PeterKietzmann** [comment]
How does the driver/device act about retransmissions and CSMA?

**Comment by PeterKietzmann** [comment]
I think setting the `page` via shell does not work correctly

**Comment by PeterKietzmann** [comment]
You implemented `NETOPT_AUTOACK`. Would it be reasonable to change to `NETOPT_ACK_REQ`? Compare #5297

File: Makefile.dep [hunk 7]
```diff
@@ -21,6 +21,15 @@ ifneq (,$(filter gnrc_netdev_default,$(USEMODULE)))
...
 
+ifneq (,$(filter cc2538_rf,$(USEMODULE)))
+  USEMODULE += netdev2_ieee802154
+  ifneq (,$(filter gnrc_netdev_default,$(USEMODULE)))
+	# XXX: this can be modelled as a dependency for gnrc_netdev_default as soon
+	# as all drivers are ported to netdev2
+    USEMODULE += gnrc_netdev2
+  endif
+endif
+
```
Comment (PeterKietzmann) [inline Makefile.dep]: Is this needed? I think the Makefile.dep in RIOT's root directory is not the correct place. It is more board- or cpu specific, right?

File: cpu/cc2538/include/cc2538_rf.h [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+    rfc_obs_sig2 = 2,
+};
+
+/* Values for RFCORE_XREG_RFC_OBS_CTRLx registers: */
+enum {
+    constant_value_0 = 0x00, /**< Constant value */
+    constant_value_1 = 0x01, /**< Constant value */
+    rfc_sniff_data   = 0x08, /**< Data from packet sniffer. Sample data on rising edges of sniff_clk. */
+    rfc_sniff_clk    = 0x09, /**< 250kHz clock for packet sniffer data. */
+    rssi_valid       = 0x0c, /**< Pin is high when the RSSI value has been updated at least once since RX was started. Cleared when leaving RX. */
```
Comment (PeterKietzmann) [inline cpu/cc2538/include/cc2538_rf.h]: I don't want to be too picky but according to our [coding conventions](https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions) we aim for no more than 80 characters per line, the absolute maximum should be 100 characters per line.

File: cpu/cc2538/radio/cc2538_rf_netdev.c [hunk 9]
```diff
@@ -0,0 +1,382 @@
...
+{
+    cc2538_rf_t *dev = (cc2538_rf_t *) netdev;
+    _dev = netdev;
+
+    mutex_lock(&dev->mutex);
+
+    uint16_t pan = cc2538_get_pan();
+    uint16_t chan = cc2538_get_chan();
+    uint16_t addr_short = cc2538_get_addr_short();
+    uint64_t addr_long = cc2538_get_addr_long();
```
Comment (PeterKietzmann) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: How about using CPUIDs?

**Comment by PeterKietzmann** [comment]
@aeneby thanks for the great work! The code looks nice, documentation exists and pinging, sending/receiving udp packets to/from an atmel node as well as setting most options worked like charm. There are just some minor comments to answer and adapt.

**Comment by PeterKietzmann** [comment]
BTW: I tested on a `remote-reva` (Zolertia RE-MOTE revision A)

**Comment by aeneby** [comment]
Thanks for the feedback @PeterKietzmann 

> Is it intended that hardware addresses (and derived by these also IPv6 addresses) are non-volatile?

The cc2538 has two non-volatile memory locations for storing persistent IEEE MACs, one writable ("secondary") and one permanent ("primary"; factory set). I was under the impression that the secondary one was being used to generate the long address automatically, but on second glance I'm not sure this is actually the case. Do you think we should be using the primary IEEE MAC as the long address until the default address topic has been fully discussed? What about the short address?

> How does the driver/device act about retransmissions and CSMA?

Not yet implemented.

> I think setting the page via shell does not work correctly

The only valid page for this radio is 0, so attempting to set it to anything else should result in an error. I see now that this is not the case, and can see why (missing `else` clause). Well spotted.

> You implemented NETOPT_AUTOACK. Would it be reasonable to change to NETOPT_ACK_REQ?

Both `NETOPT_AUTOACK` and `NETOPT_ACK_REQ` are already implemented for this driver.

> Is this needed? I think the Makefile.dep in RIOT's root directory is not the correct place.

You're right - I see now that the corresponding stanzas for the other 802.15.4 drivers have been moved out of here, and into `drivers/Makefile.dep` instead. I'll have a look at changing/removing this.

Agreed about line-length, will fix.

**Comment by PeterKietzmann** [comment]
>Do you think we should be using the primary IEEE MAC as the long address until the default address topic has been fully discussed

In case this one is "unique" over multiple devices, I think it makes sense to use it as default address. IP addresses should still be adaptable but fall back to the (same) default address as always. **If** the last bytes of the long hardware address have enough entropy, just use the last bytes of the long address. Otherwise, you could think of a quickfix as done in #5461.

>Both `NETOPT_AUTOACK` and `NETOPT_ACK_REQ` are already implemented for this driver

Where do I find  the implementation of `NETOPT_ACK_REQ` ? I've assumed this option is shown in the shell. Was that wrong?
```
Iface  7   HWaddr: 79:bc  Channel: 26  Page: 0  NID: 0x23
             Long HWaddr: 9b:df:6f:df:ff:b7:b9:7f 
             TX-Power: 3dBm  State: IDLE 
             AUTOACK  MTU:1280  HL:64  6LO  RTR  IPHC  
             Source address length: 8
             Link type: wireless
             inet6 addr: ff02::1/128  scope: local [multicast]
             inet6 addr: fe80::99df:6fdf:ffb7:b97f/64  scope: local
             inet6 addr: ff02::1:ffb7:b97f/128  scope: local [multicast]
             inet6 addr: ff02::1a/128  scope: local [multicast]
```

> Not yet implemented.

Fine with me. But would be nice to have this in near future ;-)!

Thanks for your feedback. Looking forward to get this merged before feature-freeze.

**Comment by aeneby** [comment]
> Where do I find the implementation of NETOPT_ACK_REQ ? I've assumed this option is shown in the shell. Was that wrong?

No - you should be able to enable it via the shell with `ifconfig <if_id> ack_req`.

**Comment by PeterKietzmann** [comment]
Okay, will test tomorrow. If we want to get this into the next release it would be great if you could update the PR by Thursday.

**Comment by aeneby** [comment]
Updated initial commit and rebased to current master. Summary of changes:

- Fixed setting of page number; setting to anything other than page 0 should now return an error.
- Added definitions for both primary and secondary EUI-64 address memory locations
- Introduced new function `cc2538_get_eui64_primary()` to retrieve the primary (burned-in) ID, and updated driver init function to set primary EUI-64 as long address using this function
- Removed a bunch of doc strings which were too long (this information is in the user guide anyway)
- Moved the inclusion of the `netdev2_ieee802154` module from `Makefile.dep` to `boards/openmote-cc2538/Makefile.dep`

Let me know if I have missed anything.

File: cpu/cc2538/radio/cc2538_rf_netdev.c (hunk 9 above)
Comment (aeneby) [inline cpu/cc2538/radio/cc2538_rf_netdev.c]: These variables are only used for updating the `netdev2` struct with whatever the hardware was already configured as during driver init - in this case the long address will already be set to a unique EUI-64 address.

File: boards/openmote-cc2538/Makefile.dep [hunk 10]
```diff
@@ -0,0 +1,6 @@
+ifneq (,$(filter gnrc_netdev_default,$(USEMODULE)))
+    USEMODULE += netif
+    USEMODULE += cc2538_rf
+    USEMODULE += gnrc_netdev2
+    USEMODULE += netdev2_ieee802154
```
Comment (kYc0o) [inline boards/openmote-cc2538/Makefile.dep]: This module is not missing on the other platforms?

Comment (aeneby) [inline boards/openmote-cc2538/Makefile.dep]: Nice catch - fixed.

**Comment by kYc0o** [comment]
I re-tested in remote-reva with success. Code looks ok and main functionalities are wroking. Let's see what Murdock thinks about ;)

**Comment by kYc0o** [comment]
Murdock is complaining about the doxygen doc on `cpu/cc2538/include/cc2538_rf.h`. Can you check it?

**Comment by aeneby** [comment]
Missing close-bracket - let's hope that was it...

**Comment by aeneby** [comment]
I guess Murdock didn't like that the `cc2538_rf_t` struct members were left undocumented. Trying again...

**Comment by PeterKietzmann** [comment]
Confirmed on a remote-reva again:
- default IP address
- documentation line width
- `NETOPT_ACK_REQ`
- page number
- diverse other netopts
- pinging from/to node with link local address
- udp server/send
- Murdock

**Comment by PeterKietzmann** [comment]
ACK and go

**Comment by aeneby** [comment]
Thanks everyone.

**Comment by OlegHahm** [comment]
Coooool! Congrats, very nice work (and very much appreciated)!

**Comment by alignan** [comment]
Thanks for the PR and the work done!

**Comment by hexluthor** [comment]
Great job @aeneby! Thank you for following-through with this.

**Pull request merged**